pkg/server/status.go

```diff
@@ -1989,6 +1989,104 @@ func (s *statusServer) NodeUI(
 	return &resp, nil
 }
 
+// NetworkConnectivity collects information about connections statuses across all nodes.
+func (s *systemStatusServer) NetworkConnectivity(
+	ctx context.Context, req *serverpb.NetworkConnectivityRequest,
+) (*serverpb.NetworkConnectivityResponse, error) {
+	ctx = forwardSQLIdentityThroughRPCCalls(ctx)
+	ctx = s.AnnotateCtx(ctx)
+
+	response := &serverpb.NetworkConnectivityResponse{
+		Connections:    map[roachpb.NodeID]serverpb.NetworkConnectivityResponse_Connectivity{},
+		ErrorsByNodeID: map[roachpb.NodeID]string{},
+	}
+
+	if len(req.NodeID) > 0 {
+		sourceNodeID, local, err := s.parseNodeID(req.NodeID)
+		if err != nil {
+			return nil, serverError(ctx, err)
+		}
+
+		if local {
+			peers := map[roachpb.NodeID]serverpb.NetworkConnectivityResponse_Peer{}
+			var nodeIDs []roachpb.NodeID
+			err = s.gossip.IterateInfos(gossip.KeyNodeDescPrefix, func(k string, info gossip.Info) error {
+				nodeIDs = append(nodeIDs, info.NodeID)
+				return nil
+			})
+			if err != nil {
+				return nil, serverError(ctx, err)
+			}
+
+			latencies := s.rpcCtx.RemoteClocks.AllLatencies()
+
+			for _, targetNodeId := range nodeIDs {
+				if sourceNodeID == targetNodeId {
+					continue
+				}
+				peer := serverpb.NetworkConnectivityResponse_Peer{}
+				peer.Latency = latencies[targetNodeId]
+
+				node, err := s.gossip.GetNodeDescriptor(targetNodeId)
+				if err != nil {
+					peer.Status = serverpb.NetworkConnectivityResponse_UNKNOWN
+					cause := errors.UnwrapAll(err)
+					peer.Error = cause.Error()
+					continue
+				}
+				if err = s.rpcCtx.ConnHealth(node.Address.AddressField, node.NodeID, rpc.SystemClass); err != nil {
+					if errors.Is(rpc.ErrNotHeartbeated, err) {
+						peer.Status = serverpb.NetworkConnectivityResponse_ESTABLISHING
+					} else {
+						peer.Status = serverpb.NetworkConnectivityResponse_ERROR
+					}
+					cause := errors.UnwrapAll(err)
+					peer.Error = cause.Error()
+				} else {
+					peer.Status = serverpb.NetworkConnectivityResponse_ESTABLISHED
+				}
+				peer.Address = node.Address.AddressField
+				peer.Locality = &node.Locality
+
+				peers[targetNodeId] = peer
+			}
+
+			response.Connections[sourceNodeID] = serverpb.NetworkConnectivityResponse_Connectivity{
+				Peers: peers,
+			}
+			return response, nil
+		}
+
```

Comment:
// A non-local NodeID was specified, so relay to another node.

I think this control flow would be easier (and the bulk of code less indented) if you switched it to

```go
if len(req.NodeID) > 0 {
  sourceNodeID, local, err := s.parseNodeID(req.NodeID)
  if err != nil {
	  return nil, serverError(ctx, err)
  }}
  if !local {
    // do the stuff that's below this comment
  }
  // Do the stuff that is now in `if local`.
}
```